pkg/controller/infrastructure/infraflow/context.go

```diff
@@ -101,52 +108,207 @@ const (
+	// }
 	return flowContext, nil
 }
 
+func (c *FlowContext) persistState(ctx context.Context) error {
+	return PatchProviderStatusAndState(ctx, c.runtimeClient, c.infra, nil, c.computeInfrastructureState(), c.getEgressCIDRs())
+}
+
+func PatchProviderStatusAndState(
+	ctx context.Context,
+	runtimeClient client.Client,
+	infra *extensionsv1alpha1.Infrastructure,
+	status *awsv1alpha1.InfrastructureStatus,
+	state *runtime.RawExtension,
+	egressCIDRs []string,
+) error {
+	patch := client.MergeFrom(infra.DeepCopy())
+	if status != nil {
+		infra.Status.ProviderStatus = &runtime.RawExtension{Object: status}
+		if egressCIDRs != nil {
+			infra.Status.EgressCIDRs = egressCIDRs
+		}
+	}
+
+	if state != nil {
+		infra.Status.State = state
+	}
+
+	// do not make a patch request if nothing has changed.
+	if data, err := patch.Data(infra); err != nil {
+		return fmt.Errorf("failed getting patch data for infra %s: %w", infra.Name, err)
+	} else if string(data) == `{}` {
+		return nil
+	}
+
+	return runtimeClient.Status().Patch(ctx, infra, patch)
+}
+
+func (c *FlowContext) computeInfrastructureStatus() *awsv1alpha1.InfrastructureStatus {
+	status := &awsv1alpha1.InfrastructureStatus{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: awsv1alpha1.SchemeGroupVersion.String(),
+			Kind:       "InfrastructureStatus",
+		},
+	}
+
+	vpcID := ptr.Deref(c.state.Get(IdentifierVPC), "")
+	groupID := ptr.Deref(c.state.Get(IdentifierNodesSecurityGroup), "")
+	ec2KeyName := ptr.Deref(c.state.Get(NameKeyPair), "")
+	iamInstanceProfileName := ptr.Deref(c.state.Get(NameIAMInstanceProfile), "")
+	arnIamRole := ptr.Deref(c.state.Get(ARNIAMRole), "")
+
+	if c.config.Networks.VPC.ID != nil {
+		vpcID = *c.config.Networks.VPC.ID
+	}
+
+	if vpcID != "" {
+		var subnets []awsv1alpha1.Subnet
+		prefix := ChildIdZones + shared.Separator
+		for k, v := range c.state.AsMap() {
+			if !shared.IsValidValue(v) {
+				continue
+			}
+			if strings.HasPrefix(k, prefix) {
+				parts := strings.Split(k, shared.Separator)
+				if len(parts) != 3 {
+					continue
+				}
+				var purpose string
+				switch parts[2] {
+				case IdentifierZoneSubnetPublic:
+					purpose = awsapi.PurposePublic
+				case IdentifierZoneSubnetWorkers:
+					purpose = awsapi.PurposeNodes
+				default:
+					continue
+				}
+				subnets = append(subnets, awsv1alpha1.Subnet{
+					ID:      v,
+					Purpose: purpose,
+					Zone:    parts[1],
+				})
+			}
+		}
+
+		status.VPC = awsv1alpha1.VPCStatus{
+			ID:      vpcID,
+			Subnets: subnets,
+		}
+		if groupID != "" {
+			status.VPC.SecurityGroups = []awsv1alpha1.SecurityGroup{
+				{
+					Purpose: awsapi.PurposeNodes,
+					ID:      groupID,
+				},
+			}
+		}
+	}
+
+	if ec2KeyName != "" {
+		status.EC2.KeyName = ec2KeyName
+	}
+
+	if iamInstanceProfileName != "" {
+		status.IAM.InstanceProfiles = []awsv1alpha1.InstanceProfile{
+			{
+				Purpose: awsapi.PurposeNodes,
+				Name:    iamInstanceProfileName,
+			},
+		}
+	}
+	if arnIamRole != "" {
```

Comment:
nit: we use `vpcID`, shouldn't it be `arnIAMRole` then?

feel free to disagree

Reply:
agree, changed to `arnIAMRole`